tutorials/17_vi_posteriors.ipynb

```diff
@@ -0,0 +1,361 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "source": [
+    "# Using Variational Inference for Building Posteriors"
+   ],
+   "metadata": {
+    "collapsed": false
+   }
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "In the previous tutorial, we saw how to build the posterior and how to specialize on one specific observation `x_o`. If one uses SNPE, then the posterior can be sampled from directly, yet this comes at the expense of necessary correction terms during training, since the samples are obtained from the \"wrong\" prior for `num_rounds > 1`. For SNLE or SNRE, MCMC sampling is required, which is computationally expensive. With SNVI (sequential neural variational inference), it is possible to directly sample from the posterior without any corrections during training or without expensive MCMC for sampling. This is possible by learning the posterior with variational inference techniques. For this, an additional network (one for the likelihood or likelihood-to-evidence-ratio) must be trained first."
+   ],
+   "metadata": {
+    "collapsed": false
+   }
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "## Main syntax"
+   ],
+   "metadata": {
+    "collapsed": false
+   }
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "outputs": [],
+   "source": [
+    "inference = SNLE(prior)\n",
+    "\n",
+    "for _ in range(num_rounds):\n",
+    "    theta, x = simulate_for_sbi(simulator, proposal, num_simulations=500)\n",
+    "\n",
+    "    # In `SNLE` and `SNRE`, you should not pass the `proposal` to `\n",
+    "    # .append_simulations()`.\n",
+    "    likelihood_estimator = inference.append_simulations(\n",
+    "        theta, x,\n",
+    "    ).train()\n",
+    "\n",
+    "    # Obtain potential (learned likelihood * prior) and theta transformation.\n",
+    "    potential_fn, theta_transform = likelihood_estimator_based_potential(\n",
+    "        likelihood_estimator, prior, x_o\n",
+    "    )\n",
+    "\n",
+    "    # Build posterior via variational inference.\n",
+    "    posterior = VIPosterior(\n",
+    "        potential_fn, prior, \"maf\", theta_transform, vi_method=\"fKL\",\n",
+    "    ).train()\n",
+    "    proposal = posterior"
```

Comment:
This line is suboptimal, as the variational distribution is recreated from scratch every round. It would be more efficient to initialize with the previous estimate in consecutive rounds.

```python
# Build posterior via variational inference.
posterior = VIPosterior(
    potential_fn, prior, "maf", theta_transform, vi_method="fKL",   # This initializes all from scratch!
).train()
proposal = posterior
```
Instead use something like this:
```python
# Build posterior via variational inference.
if num_rounds == 0:
  posterior = VIPosterior(
      potential_fn, prior, "maf", theta_transform, vi_method="fKL",   # This initializes all from scratch!
  ).train()
else:
  posterior = VIPosterior(
        potential_fn, prior, proposal, theta_transform, vi_method="fKL",   # This initializes all from scratch!
    ).train()
proposal = posterior
```

Reply:
Hi, this doesn't work, here the truncated stack trace: 

´´´
AssertionError                            Traceback (most recent call last)
Cell In[6], line 21
     17     posterior = VIPosterior(
     18         potential_fn, prior, "maf", theta_transform, vi_method="fKL",
     19     ).train()
     20 else: 
---> 21     posterior = VIPosterior(
     22         potential_fn, prior, proposal, theta_transform, vi_method="fKL",
     23     ).train()
     24 posteriors.append(posterior)
     25 proposal = posterior

File [~/anaconda3/envs/sbi_env/lib/python3.11/site-packages/sbi/inference/posteriors/vi_posterior.py:131](https://file+.vscode-resource.vscode-cdn.net/home/imahn/sbi/tutorials/~/anaconda3/envs/sbi_env/lib/python3.11/site-packages/sbi/inference/posteriors/vi_posterior.py:131), in VIPosterior.__init__(self, potential_fn, prior, q, theta_transform, vi_method, device, x_shape, parameters, modules)
    128     self.link_transform = theta_transform.inv
    130 # This will set the variational distribution and VI method
--> 131 self.set_q(q, parameters=parameters, modules=modules)
    132 self.set_vi_method(vi_method)
    134 self._purpose = (
    135     "It provides Variational inference to .sample() from the posterior and "
    136     "can evaluate the _normalized_ posterior density with .log_prob()."
    137 )

File [~/anaconda3/envs/sbi_env/lib/python3.11/site-packages/sbi/inference/posteriors/vi_posterior.py:247](https://file+.vscode-resource.vscode-cdn.net/home/imahn/sbi/tutorials/~/anaconda3/envs/sbi_env/lib/python3.11/site-packages/sbi/inference/posteriors/vi_posterior.py:247), in VIPosterior.set_q(self, q, parameters, modules)
    243 move_all_tensor_to_device(q, self._device)
...
    115         an iterable of parameters."""
    116     )

AssertionError: Nothing to train, atleast one of the parameters must have an enabled
            gradient.
´´´